File: pipelines/leaderboard_scoring.py
"""
Mirror lib/server/leaderboard-scoring.js for salon_ai_leaderboard ingestion.

Scoring v2 (2026): see leaderboard-scoring.js — evidence-based ai_score,
histogram-mean sentiment. assessment_level in DB uses global cohort percentiles
(latest ∪ batch) when ingesting via apply_global_percentile_assessment_levels;
build_assessment_level remains as a fixed-threshold fallback for tooling.
"""

from __future__ import annotations

import math
import re
import time
from typing import Any, Mapping

SLUG_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


def empty_histogram() -> dict[int, int]:
    return {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}


def sum_histogram(hist: Mapping[int, Any] | None) -> int:
    h = hist or {}
    return sum(int(h.get(s, 0) or 0) for s in range(1, 6))


def normalize_histogram_input(raw: Any) -> dict[int, int] | None:
    if not isinstance(raw, dict):
        return None
    hist = empty_histogram()
    any_val = False
    for s in range(1, 6):
        v = raw.get(str(s), raw.get(s))
        if v is not None and v != "":
            try:
                n = int(float(v))
            except (TypeError, ValueError):
                continue
            if n >= 0:
                hist[s] = n
                any_val = True
    return hist if any_val else None


def count_stars_from_review_list(reviews: Any) -> dict[int, int]:
    hist = empty_histogram()
    if not isinstance(reviews, list):
        return hist
    for rv in reviews:
        if not isinstance(rv, dict):
            continue
        try:
            rating = float(rv.get("rating"))
        except (TypeError, ValueError):
            continue
        if not math.isfinite(rating):
            continue
        rounded = int(round(rating))
        if 1 <= rounded <= 5:
            hist[rounded] += 1
    return hist


def classify_store_signals(
    histogram: Mapping[int, Any] | None,
    meta: dict[str, Any] | None,
) -> dict[str, Any]:
    hist = dict(empty_histogram())
    if histogram:
        for s in range(1, 6):
            hist[s] = int(histogram.get(s, 0) or 0)
    c1 = hist[1]
    c2 = hist[2]
    total = sum_histogram(hist)
    incomplete = bool(meta and meta.get("histogramIncomplete"))
    total_on_google = meta.get("userRatingTotal") if meta else None
    try:
        total_on_google_f = float(total_on_google) if total_on_google is not None else None
    except (TypeError, ValueError):
        total_on_google_f = None

    base: dict[str, Any] = {
        "qualityTier": None,
        "redAlert": False,
        "histogramIncomplete": incomplete,
    }

    if total == 0:
        base["qualityTier"] = "unknown"
        return base

    if c1 >= 3:
        return {**base, "redAlert": True, "qualityTier": "alert"}
    if c1 == 2:
        return {**base, "redAlert": True, "qualityTier": "alert"}
    if c1 == 1:
        return {**base, "redAlert": True, "qualityTier": "alert"}
    if c2 >= 1:
        return {**base, "redAlert": True, "qualityTier": "alert"}

    present = [s for s in range(1, 6) if hist[s] > 0]
    min_star = min(present)
    max_star = max(present)

    quality_tier = "mixed"
    if min_star == 5 and max_star == 5:
        quality_tier = "excellent"
    elif min_star >= 4:
        quality_tier = "good"
    elif min_star >= 3:
        quality_tier = "moderate"

    out = {**base, "qualityTier": quality_tier, "redAlert": False}
    if incomplete and total_on_google_f is not None and total_on_google_f > total:
        out["notes"] = "histogram sample smaller than Google total"
    return out


def sentiment_from_review_histogram(histogram: Mapping[int, Any] | None, user_rating_total: int) -> float:
    meta = {"userRatingTotal": user_rating_total, "histogramIncomplete": True}
    sig = classify_store_signals(histogram, meta)
    if sig.get("redAlert"):
        return 0.58

    hist = dict(histogram or {})
    total = 0
    wsum = 0.0
    for s in range(1, 6):
        c = int(hist.get(s, 0) or 0)
        total += c
        wsum += float(s) * c
    if total <= 0:
        return 0.8

    avg_star = wsum / total
    linear = (avg_star - 1.0) / 4.0
    return round(min(0.98, max(0.45, linear)) * 1000) / 1000


def calc_ai_score(rating: float, review_count: int, sentiment_p: float, freshness_f: float) -> float:
    r = float(rating) if rating is not None else 0.0
    r = min(5.0, max(1.0, r))
    n = max(0, int(review_count or 0))
    pn = sentiment_p if math.isfinite(sentiment_p) else 0.72
    pn = min(1.0, max(0.0, pn))
    fn = freshness_f if math.isfinite(freshness_f) else 0.7
    fn = min(1.0, max(0.0, fn))

    rating_norm = min(1.0, max(0.0, (r - 3.0) / 2.0))
    volume_norm = min(1.0, math.log10(n + 1) / math.log10(900))
    blend = rating_norm * 0.31 + volume_norm * 0.33 + pn * 0.22 + fn * 0.14
    conf = 1.0 - math.exp(-n / 52.0)
    prior = 0.52
    adjusted = prior + (blend - prior) * (0.36 + 0.64 * conf)
    out = adjusted * 100.0
    return round(min(100.0, max(0.0, out)) * 10) / 10


def assessment_level_from_fraction_from_top(frac: float) -> str:
    """0 = best in cohort. Bands: 10% / 20% / 30% / 20% / 20% (GOOD 10–30%, fair MODERATE 30–60%)."""
    if not math.isfinite(frac) or frac < 0:
        return "MODERATE"
    if frac < 0.1:
        return "EXCELLENT"
    if frac < 0.3:
        return "GOOD"
    if frac < 0.6:
        return "MODERATE"
    if frac < 0.8:
        return "LOW"
    return "RISKY"


def assign_assessment_levels_by_ai_score_percentile(rows: list[dict[str, Any]]) -> None:
    """Set assessment_level by ai_score rank within ``rows`` (desc). Mutates rows."""
    triples = [(i, str(r.get("slug") or "").strip(), float(r.get("ai_score") or 0)) for i, r in enumerate(rows)]
    triples = [t for t in triples if t[1]]
    triples.sort(key=lambda t: (-t[2], t[1]))
    n = len(triples)
    if n == 0:
        return
    for k, (orig_i, _slug, _sc) in enumerate(triples):
        frac = 0.0 if n == 1 else k / n
        rows[orig_i]["assessment_level"] = assessment_level_from_fraction_from_top(frac)


def apply_global_percentile_assessment_levels(client: Any, rows: list[dict[str, Any]]) -> None:
    """Merge DB latest listed scores with this batch; set each row's assessment_level."""
    resp = client.table("salon_ai_leaderboard_latest").select("slug", "ai_score").execute()
    merged: dict[str, float] = {}
    for item in resp.data or []:
        s = str(item.get("slug") or "").strip()
        if s:
            merged[s] = float(item.get("ai_score") or 0)
    for r in rows:
        s = str(r.get("slug") or "").strip()
        if s:
            merged[s] = float(r.get("ai_score") or 0)
    ordered = sorted(merged.keys(), key=lambda slug: (-merged[slug], slug))
    n = len(ordered)
    slug_level: dict[str, str] = {}
    for k, slug in enumerate(ordered):
        frac = 0.0 if n == 1 else k / n
        slug_level[slug] = assessment_level_from_fraction_from_top(frac)
    for r in rows:
        s = str(r.get("slug") or "").strip()
        if s:
            r["assessment_level"] = slug_level.get(s, "MODERATE")


def build_assessment_level(rating: float, score: float) -> str:
    """Legacy fixed thresholds (rating + ai_score). Prefer percentile helpers for DB."""
    r = float(rating) if rating is not None else 0.0
    s = float(score) if score is not None else 0.0
    if s >= 86 and r >= 4.71:
        return "EXCELLENT"
    if s >= 70 and r >= 4.36:
        return "GOOD"
    if s >= 53 and r >= 3.98:
        return "MODERATE"
    if r < 4.0:
        return "LOW"
    return "RISKY"


def freshness_heuristic(review_count: int, reviews_array: list[Any] | None) -> float:
    n = max(0, int(review_count or 0))
    recent = 0
    total_with_time = 0
    if isinstance(reviews_array, list):
        now_ms = time.time() * 1000
        max_age_ms = 120 * 24 * 60 * 60 * 1000
        for rv in reviews_array:
            if not isinstance(rv, dict):
                continue
            pt = rv.get("publishTime") or rv.get("publish_time")
            if not pt:
                continue
            total_with_time += 1
            try:
                t_ms = _parse_time_ms(str(pt))
            except (ValueError, OSError, OverflowError):
                continue
            if math.isfinite(t_ms) and now_ms - t_ms <= max_age_ms:
                recent += 1
    if total_with_time >= 3:
        ratio = recent / total_with_time
        return round((0.64 + ratio * 0.34) * 1000) / 1000
    return round((0.58 + min(1.0, n / 500.0) * 0.36) * 1000) / 1000


def _parse_time_ms(iso: str) -> float:
    from datetime import datetime

    s = iso.strip()
    if s.endswith("Z"):
        s = s[:-1] + "+00:00"
    dt = datetime.fromisoformat(s)
    return dt.timestamp() * 1000


def sentiment_fallback_from_rating(rating: float) -> float:
    r = float(rating) if rating is not None else 0.0
    r = min(5.0, max(1.0, r))
    linear = (r - 1.0) / 4.0
    return round(min(0.98, max(0.52, linear)) * 1000) / 1000


def clamp01(x: float) -> float:
    if not math.isfinite(x):
        return 0.72
    return min(1.0, max(0.0, x))


def histogram_from_record(rec: Mapping[str, Any]) -> dict[int, int]:
    if isinstance(rec.get("reviews"), list):
        return count_stars_from_review_list(rec["reviews"])
    raw_hist = rec.get("star_histogram") or rec.get("starHistogram") or rec.get("histogram")
    normalized = normalize_histogram_input(raw_hist)
    return normalized if normalized is not None else empty_histogram()


def compute_leaderboard_metrics(
    rating: float,
    review_count: int,
    reviews: list[Any] | None,
    star_histogram_raw: Any,
) -> dict[str, Any]:
    """Build metrics from one logical source row (rating, n, optional reviews / histogram keys on parent)."""
    rec: dict[str, Any] = {
        "reviews": reviews,
        "star_histogram": star_histogram_raw,
    }
    hist = histogram_from_record(rec)
    hist_total = sum_histogram(hist)
    incomplete = hist_total > 0 and review_count > 0 and hist_total < review_count
    if hist_total > 0:
        sentiment_p = sentiment_from_review_histogram(hist, review_count or hist_total)
        if incomplete:
            blend = clamp01((sentiment_p + sentiment_fallback_from_rating(rating)) / 2)
            sentiment_p = round(blend * 1000) / 1000
    else:
        sentiment_p = sentiment_fallback_from_rating(rating)

    reviews_arr = reviews if isinstance(reviews, list) else []
    freshness_f = freshness_heuristic(review_count, reviews_arr)
    ai_score = calc_ai_score(rating, review_count, sentiment_p, freshness_f)
    level = build_assessment_level(rating, ai_score)
    return {
        "sentiment_p": round(sentiment_p * 1000) / 1000,
        "freshness_f": freshness_f,
        "ai_score": ai_score,
        "assessment_level": level,
    }


def ensure_slug(raw_slug: str | None, name: str, index: int) -> str:
    trimmed = (raw_slug or "").strip().lower()
    if trimmed and SLUG_RE.fullmatch(trimmed):
        return trimmed
    base = re.sub(r"[^a-z0-9]+", "-", (name or "salon").lower()).strip("-")[:40] or "salon"
    candidate = f"{base}-{index}"
    if SLUG_RE.fullmatch(candidate):
        return candidate
    return f"salon-{index}"

File: lib/server/leaderboard-scoring.js
/**
 * AI leaderboard scoring (salon_ai_leaderboard).
 *
 * Design goals (v2, 2026):
 * - Spread salons in the typical 4.2–4.9★ / tens–hundreds of reviews band instead of clustering near 90+.
 * - Treat review volume as evidence: few reviews → score pulled toward a neutral prior (Bayesian-style shrink).
 * - Sentiment_p from star histogram uses the sample mean star (continuous), not only coarse tiers.
 * - Freshness still rewards recent review timestamps when present; without dates, volume-only curve is slightly flatter.
 * - assessment_level is cohort-relative on latest listed rows: rank by ai_score
 *   (desc); fraction from the top → EXCELLENT (top 10%), GOOD (10–30%),
 *   MODERATE / “fair” (30–60%), LOW / “poor” (60–80%), RISKY / “critical” (80–100%).
 * - buildAssessment(rating, score) remains for legacy / tooling; live ingest & refresh use percentiles.
 *
 * Keep in sync: pipelines/leaderboard_scoring.py
 */

const { classifyStoreSignals, sumHistogram } = require("./store-map-report");

function clamp01(x, fallback) {
  const n = Number(x);
  if (!Number.isFinite(n)) return fallback;
  return Math.min(1, Math.max(0, n));
}

/**
 * Composite 0–100 AI index from Google-style signals.
 * @param {number} rating       Average stars (1–5)
 * @param {number} reviewCount  userRatingCount-style total
 * @param {number} sentimentP   0–1 reputation from reviews/histogram
 * @param {number} freshnessF   0–1 recency heuristic
 */
function calcAiScore(rating, reviewCount, sentimentP, freshnessF) {
  const r = Math.min(5, Math.max(1, Number(rating) || 0));
  const n = Math.max(0, Math.floor(Number(reviewCount) || 0));
  const pn = clamp01(sentimentP, 0.72);
  const fn = clamp01(freshnessF, 0.7);

  // 3★ → 0, 5★ → 1 (typical salons sit mid-band, not automatically ~0.9+)
  const ratingNorm = Math.min(1, Math.max(0, (r - 3) / 2));
  // Volume plateaus around ~900 reviews (log dampening)
  const volumeNorm = Math.min(1, Math.log10(n + 1) / Math.log10(900));

  const blend =
    ratingNorm * 0.31 + volumeNorm * 0.33 + pn * 0.22 + fn * 0.14;

  // Evidence: asymptotic confidence from count (~half weight at ~40 reviews, ~90% at ~120)
  const conf = 1 - Math.exp(-n / 52);
  const prior = 0.52;
  const adjusted = prior + (blend - prior) * (0.36 + 0.64 * conf);
  const out = adjusted * 100;
  return Math.round(Math.min(100, Math.max(0, out)) * 10) / 10;
}

/**
 * @param {number} frac 0 = best salon in cohort, 1 = worst (exclusive upper bands use k/n).
 * @returns {"EXCELLENT"|"GOOD"|"MODERATE"|"LOW"|"RISKY"}
 */
function assessmentLevelFromFractionFromTop(frac) {
  const f = Number(frac);
  if (!Number.isFinite(f) || f < 0) return "MODERATE";
  if (f < 0.1) return "EXCELLENT";
  if (f < 0.3) return "GOOD";
  if (f < 0.6) return "MODERATE";
  if (f < 0.8) return "LOW";
  return "RISKY";
}

/**
 * Assign assessment_level from ai_score rank within this array (best score first).
 * Mutates each row.assessment_level; rows need slug + ai_score (number).
 * @param {{ slug?: string, ai_score?: number, assessment_level?: string }[]} rows
 */
function assignAssessmentLevelsByAiScorePercentile(rows) {
  if (!Array.isArray(rows) || rows.length === 0) return;
  const list = rows
    .map(function (r, i) {
      return {
        i,
        slug: String((r && r.slug) || ""),
        score: Number(r && r.ai_score) || 0,
      };
    })
    .filter(function (x) {
      return x.slug;
    });
  list.sort(function (a, b) {
    if (b.score !== a.score) return b.score - a.score;
    return a.slug.localeCompare(b.slug);
  });
  const n = list.length;
  for (let k = 0; k < n; k += 1) {
    const frac = n === 1 ? 0 : k / n;
    rows[list[k].i].assessment_level = assessmentLevelFromFractionFromTop(frac);
  }
}

/**
 * mergedSlugToScore: slug -> ai_score (incoming row should overwrite same slug).
 * @param {Map<string, number>} mergedSlugToScore
 * @param {string} slug
 */
function resolveAssessmentLevelForSlug(mergedSlugToScore, slug) {
  const key = String(slug || "").trim();
  if (!key || !(mergedSlugToScore instanceof Map)) return "MODERATE";
  const entries = Array.from(mergedSlugToScore.entries()).sort(function (a, b) {
    if (b[1] !== a[1]) return b[1] - a[1];
    return a[0].localeCompare(b[0]);
  });
  const n = entries.length;
  for (let k = 0; k < n; k += 1) {
    if (entries[k][0] === key) {
      const frac = n === 1 ? 0 : k / n;
      return assessmentLevelFromFractionFromTop(frac);
    }
  }
  return "MODERATE";
}

/**
 * Risk / quality band from headline rating + composite score (v2 thresholds).
 * Prefer assessmentLevelFromFractionFromTop / assignAssessmentLevelsByAiScorePercentile for DB rows.
 */
function buildAssessment(rating, score) {
  const r = Number(rating) || 0;
  const s = Number(score) || 0;
  if (s >= 86 && r >= 4.71) {
    return {
      level: "EXCELLENT",
      emoji: "⭐",
      color: "#1A365D",
      light: "#EBF8FF",
      mid: "#4299E1",
      border: "#63B3ED",
    };
  }
  if (s >= 70 && r >= 4.36) {
    return {
      level: "GOOD",
      emoji: "🟢",
      color: "#276749",
      light: "#F0FFF4",
      mid: "#68D391",
      border: "#68D391",
    };
  }
  if (s >= 53 && r >= 3.98) {
    return {
      level: "MODERATE",
      emoji: "🟡",
      color: "#975A16",
      light: "#FFFFF0",
      mid: "#F6E05E",
      border: "#ECC94B",
    };
  }
  if (r < 4.0) {
    return {
      level: "LOW",
      emoji: "🟠",
      color: "#C05621",
      light: "#FFFAF0",
      mid: "#F6AD55",
      border: "#F6AD55",
    };
  }
  return {
    level: "RISKY",
    emoji: "🔴",
    color: "#C53030",
    light: "#FFF5F5",
    mid: "#FC8181",
    border: "#FC8181",
  };
}

/**
 * 0–1 sentiment from per-star histogram (or red-alert cap).
 */
function sentimentFromReviewHistogram(histogram, userRatingTotal) {
  const meta = { userRatingTotal: userRatingTotal, histogramIncomplete: true };
  const sig = classifyStoreSignals(histogram, meta);
  if (sig.redAlert) return 0.58;

  const hist = histogram && typeof histogram === "object" ? histogram : {};
  let total = 0;
  let wsum = 0;
  for (let star = 1; star <= 5; star += 1) {
    const c = Number(hist[star]) || 0;
    total += c;
    wsum += star * c;
  }
  if (total <= 0) return 0.8;

  const avgStar = wsum / total;
  const linear = (avgStar - 1) / 4;
  return Math.round(Math.min(0.98, Math.max(0.45, linear)) * 1000) / 1000;
}

function freshnessHeuristic(reviewCount, reviewsArray) {
  const n = Math.max(0, Number(reviewCount) || 0);
  let recent = 0;
  let total = 0;
  if (Array.isArray(reviewsArray)) {
    const now = Date.now();
    const maxAgeMs = 120 * 24 * 60 * 60 * 1000;
    reviewsArray.forEach(function (rv) {
      const pt = rv && (rv.publishTime || rv.publish_time);
      if (!pt) return;
      total += 1;
      const t = Date.parse(String(pt));
      if (Number.isFinite(t) && now - t <= maxAgeMs) recent += 1;
    });
  }
  if (total >= 3) {
    const ratio = recent / total;
    return Math.round((0.64 + ratio * 0.34) * 1000) / 1000;
  }
  return Math.round((0.58 + Math.min(1, n / 500) * 0.36) * 1000) / 1000;
}

module.exports = {
  assessmentLevelFromFractionFromTop,
  assignAssessmentLevelsByAiScorePercentile,
  buildAssessment,
  calcAiScore,
  freshnessHeuristic,
  resolveAssessmentLevelForSlug,
  sentimentFromReviewHistogram,
};
